test/benchmark/tools.py

```diff
@@ -0,0 +1,148 @@
+    error2 = error1.kron(error1)
+    noise_model.add_all_qubit_quantum_error(error2, ['cx'])
+    return noise_model
+
+
+def quantum_volume_circuit(num_qubits, depth, measure=True, seed=None):
+    """Create a quantum volume circuit without measurement.
+
+    The model circuits consist of layers of Haar random
+    elements of SU(4) applied between corresponding pairs
+    of qubits in a random bipartition.
+
+    Args:
+        num_qubits (int): number of qubits
+        depth (int): ideal depth of each model circuit (over SU(4))
+        measure (bool): include measurement in circuit.
+        seed (int): the seed for the random number generator
+
+    Returns:
+        QuantumCircuit: A quantum volume circuit.
+    """
+    # Create random number generator with possibly fixed seed
+    rng = random.RandomState(seed)
+    # Create quantum/classical registers of size n
+    qr = QuantumRegister(num_qubits)
+    circuit = QuantumCircuit(qr)
+    # For each layer
+    for _ in repeat(None, depth):
+        # Generate uniformly random permutation Pj of [0...n-1]
+        perm = rng.permutation(num_qubits)
+        # For each consecutive pair in Pj, generate Haar random SU(4)
+        # Decompose each SU(4) into CNOT + SU(2) and add to Ci
+        for k in range(math.floor(num_qubits / 2)):
+            # Generate random SU(4) matrix
+            X = (rng.randn(4, 4) + 1j * rng.randn(4, 4))
+            SU4, _ = linalg.qr(X)           # Q is a unitary matrix
+            SU4 /= pow(linalg.det(SU4), 1 / 4)   # make Q a special unitary
+            decomposed_SU4 = two_qubit_kak(SU4)  # Decompose into CX and U gates
+            qubits = [int(perm[2 * k]), int(perm[2 * k + 1])]
+            for gate in decomposed_SU4:
+                i0 = qubits[gate["args"][0]]
+                if gate["name"] == "cx":
+                    i1 = qubits[gate["args"][1]]
+                    circuit.cx(qr[i0], qr[i1])
+                elif gate["name"] == "u1":
+                    circuit.u1(gate["params"][2], qr[i0])
+                elif gate["name"] == "u2":
+                    circuit.u2(gate["params"][1], gate["params"][2], qr[i0])
+                elif gate["name"] == "u3":
+                    circuit.u3(gate["params"][0], gate["params"][1],
+                               gate["params"][2], qr[i0])
+                elif gate["name"] == "id":
+                    pass
+    if measure is True:
+        circuit = _add_measurements(circuit, qr)
+    return circuit
+
+
+def simple_u3_circuit(num_qubits, measure=True):
+    """Creates a simple circuit composed by u3 gates, with measurements or not
+    at the end of each qubit.
+
+    Args:
+        num_qubits (int): Number of qubits
+        measure (bool): Add measurements at the end of each qubit
+
+    Returns:
+        QuantumCircuit: The simple quantum circuit
+    """
+    qr = QuantumRegister(num_qubits)
+    circuit = QuantumCircuit(qr)
+    for i in range(num_qubits):
```

Comment:
It would be good to time a *single* u3 gate vs qubit number, and average over which qubit in circuit is applied to.

I.e: Estimate time `t[j]` of applying `circuit.u3(qr[j])`, and then compute `t_ave = mean(t)` and then plot `t_ave` vs total number of qubits in the circuit. 

This should also be done for `cx` (averaged overall pairs of qubits) `reset` and `measure` (averaged over single qubits), and later we could add other specialized gates that have an optimised implementation.

This would allow estimating the single-shot run-time as if we know for n-qubits what `t_ave_u3, t_ave_cx, t_ave_reset, t_ave_measure` we can just multiply them by the number of instructions of each type in the circuit.

Reply:
Ok!
Let's move this to a follow up PR :)